### ISSUE TITLE
Move Content Live

### ISSUE THREAD

**Issue by Snugug** [punchcard-cms/punchcard#136]
In order to have content available in the `live` database for APIs, it needs to be moved there

**Definitions**
* Now - The current date and time
* Live - Publicly available in the API and the `live` database the API draws from
* Sunrise - The date, time, and timezone content is scheduled to go live
* Sunset - The date, time, and timezone content is schedule to stop being live
* Approved- Content that has gone through all steps in its [workflow](https://github.com/punchcard-cms/punchcard/pull/105)

```gherkin
Feature: Publish Content
  As a content creator
  I want my approved revisions to be scheduled to go live and come down
  So that the most up-to-date and accurate 

  Scenario: Schedule Sunrise
    Given that a content revision I have created has been approved
      And that its sunrise is later than now
     When now is equal to the revision's sunrise
     Then it should replace the live revision of its content
      And if there are multiple approved revisions with the same sunrise, only the most recently approved should be made live

  Scenario: Immediate Sunrise
    Given that a content revision I have created has a sunrise
     When the content revision is approved
      And its sunrise is now or earlier
     Then it should replace the live revision

  Scenario: Schedule Sunset
    Given that a content revision I have created has been approved
      And that it has a sunset
      And that its sunset is later than now
     When now is equal to the revision's sunset
     Then the content it is a revision for should be removed from being live

  Scenario: Change Sunset
    Given that a content revision I have created has a sunset
      And the current live revision of its content also has a sunset
     When the content revision is approved
     Then the content it is a revision of should have its sunset changed to the approved revision's sunset

  Scenario: Cancel Sunset
    Given that a content revision I have created does not have a sunset
      And the current live revision of its content does
     When the content revision is approved
     Then the content it is a revision of should have its sunset removed

  Scenario: Immediate Sunset
    Given that a content revision I have created has a sunset
     When the content revision is approved
      And its sunset is now or earlier
     Then the content it is a revision for should be removed from being live


Feature: Approve Content
  As a content approver
  I want to understand how the content I'm reviewing is going to be affected by sunrise/sunset
  So that I can make an informed decision about whether or not I should approve or reject the revision being reviewed

  Scenario: Cancel Sunset
    Given that I am reviewing a content revision without a sunset
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has not yet occurred
     Then I should be notified on the approval page that approving this revision will cancel the existing sunset
      And I should be notified of the existing sunset

  Scenario: Change Sunset
    Given that I am reviewing a content revision with a sunset
      And that revision's sunset is past now
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has not yet occurred
     Then I should be notified on the approval page that this revision will change the existing sunset
      And I should be notified of the extent of the change (how much further/closer to now the change is)

  Scenario: Immediate Sunset
    Given that I am reviewing a content revision with a sunset
      And that revision's sunset is before now
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has not yet occurred
     Then I should be notified on the approval page that this revision will immediately sunset its content 

  Scenario: Sunrise That Which Is Sunset
    Given that I am reviewing a content revision with a sunrise
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has already occurred
     Then I should be notified on the approval page that this revision will sunrise the currently sunset content

  Scenario: Sunrise That Which Will Be Sunset
    Given that I am reviewing a content revision with a sunrise
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has not yet occurred
      And the sunset will occur before the current revision's sunrise
     Then I should be notified on the approval page that this revision will sunrise the content that otherwise would be sunset
      And I should be notified of the difference between when the content would be scheduled to sunset and sunrise

  Scenario: Immediate Sunrise
    Given that I am reviewing a content revision with a sunrise
     When that revision's sunrise is equal to or before now
     Then I should be notified on the approval page that this revision will be live immediately after approval

  Scenario: Sunrise Override
    Given that I am reviewing a content revision with a sunrise
     When that revision's content has existing revisions scheduled for sunrise
      And this revision's sunrise is the same as other revisions that are scheduled for sunrise
     Then I should be notified on the approval page that this revision will supersede the other scheduled revisions with the same sunrise

  Scenario: View Sunrise and Sunset
    Given that I am reviewing a content revision
     When I want to check the sunrise or sunset value of the revision being reviewed
     Then I should be able to readily see them just as I can the other fields of a revision

  Scenario: View Revision Schedule
    Given that a piece of content can have multiple revisions scheduled to sunrise at once
      And a piece of content may be scheduled to sunset
      And a piece of content my have a live revision
     When I want to understand how a piece of content may change over time
      And I can see what revision is current live, if any
     Then I can see a listing of approved revisions for a piece of content
      And I can see the revision ID, date created, date approved, identifier, have a link to view the complete revision, and have a link to create a new revision off of the current one, for each revision
      And if the revision has not yet passed its sunrise, its sunrise date and time until sunrise
      And revisions should be sorted by when they will become live
      And the current live revision should be separated from all other revisions
      And the sunset date for the content should be visible
      And if there are revisions scheduled to become live after the sunset date, they should be separated from revisions scheduled to become live before the sunset date.
      And I should be able to access this view from the approval page
      And I should be able to access this view from the revision listing page
      And I should be able to access this from the content listing page
```

---

**Publish Content**
- [ ] Schedule Sunrise
- [ ] Immediate Sunrise
- [ ] Schedule Sunset
- [ ] Change Sunset
- [ ] Cancel Sunset
- [ ] Immediate Sunset

**Approve Content**
- [ ] Cancel Sunset
- [ ] Change Sunset
- [ ] Immediate Sunset
- [ ] Sunrise That Which Is Sunset
- [ ] Sunrise That Which Will Be Sunset
- [ ] Immediate Sunrise
- [ ] Sunrise Override
- [ ] View Sunrise and Sunset
- [ ] View Revision Schedule

**Comment by scottnath** [comment]
@Snugug 

could there be another _and_ on these stating that there is only _one_ revision available from any piece of content at a time?

**Comment by Snugug** [comment]
I was thinking about that, I'm not sure I want to only have one sunrise revision available at a time. Someone may want to queue up a few different changes (for instance, homepage changes for World of Watson) and I'd want them to have those revisions queued in advanced

The main thing I'm looking for input in to is around sunset. If a revision is approved and has a sunset date, and another revision gets approved later without a sunset date, do we remove the sunset from the earlier revision? Do we message about about that to the user and take something down then put it back up? What if two revisions have a sunset date, does the latest override the current?

**Comment by SirSpidey** [comment]
@Snugug, How does this flow fit with the flow we need to have reviewable content that's not ready for publishing [...Enter content that is not ready for release so that I and others... can review](https://github.ibm.com/Watson/wox-cms/issues/97)

Seems that we need another state besides `live`. 
And that seems at odds with @scottnath 's comment that only one revision is available at a time (although this isn't really a revision, but a "branch").

And you wrote,
>If a revision is approved and has a sunset date, and another revision gets approved later without a sunset date, do we remove the sunset from the earlier revision? Do we message about about that to the user and take something down then put it back up? 

At the API/Service level for docs and API ref, we generally don't have sunset dates on _pieces_ of content; we usually are sunsetting the service. For example, we'll create a deprecation notice that has a sunrise date, then we have a sunset date for the entire service. 

>What if two revisions have a sunset date, does the latest override the current?

We have this issue currently (latest overrides current), and this causes problems because people aren't aware of the changes they have overwritten. So I think we need a review/message process of the changes. 

If 2 revisions have a sunset date, latest overrides (with clear info message of what's happening).

**Comment by Snugug** [comment]
@SirSpidey This story is specifically about the the expectations of when content is ready to move live, not about the review process before.

Currently a content editor can save a revision without pushing it through the workflow process to publish it live. Those revisions are available to review, and the workflow process exists specifically to allow individuals to view content and approve it for publishing.

As part of our [API Planning](https://github.com/punchcard-cms/punchcard/wiki/Architecture---API), Punchcard will also wind up providing a separate API under which all content will be available (with separate authentication) to allow in-application preview of unpublished content. I realize we didn't have a story to track that previously, so I've added https://github.com/punchcard-cms/punchcard/issues/272

---

So, here are my thoughts:

* Allow an unlimited number of revisions to be in the sunrise queue at once
* Allow only a single sunset date per content type
* Latest approved revision owns the single sunset date
* If a revision is in the review process, and there is a sunset date for the content UUID already in place, and the sunset date is different than the one on the current revision, message the approver to notify them that the current revision will change the sunset date.
  * If there is no sunset date on the revision being reviewed, message the reviewer that the sunset date is being removed and the content will not be sunset
  * If there is a sunset date on the revision being reviewed, and it's past now, and before the existing sunset date, message the reviewer that they are sunsetting the content earlier than before
  * If there is a sunset date on the revision being reviewed, and its past now, and after the existing sunset date, message the reviewer that they are extending the content's lifespan
  * If there is a sunset date on the revision being reviewed, and it's before now, message the reviewer that they are about to take the content down
  * If a revision is being reviewed, and the sunrise date of the revision is past the current sunset date, or the content has already been sunset, message the reviewer that that content will become available again

---

While out-of-scope for this issue, if messaging referenced content that a referencer is being sunset, a story should be written and added to the [Reference Epic](https://github.com/punchcard-cms/punchcard/issues/198)

**Comment by scottnath** [comment]
`Do we message about about that to the user`

The `user` here is the `application developer`, correct?
(this should be Core Developer...or the [personas](https://github.com/punchcard-cms/punchcard/wiki/Personas) should be changed)

So, for the application developer, I can see the point of the various messaging for what is/is not live. 

*Why is messaging not part of the scenarios?*

`Allow only a single sunset date per content type`

I'm against having a common sunset date. That severely limits the usefulness of sunset. Each revision should be able to have their own sunrise and sunset dates.

```gherkin
  Scenario: New Content takes over
    Given that I have a new revision of an existing piece of content
      And that the current live revision should be taken down at a scheduled time in the future
    When I want the new revision to supersede the existing revision
    Then I should select a sunset date for the live revision
      And that sunset date should bring the live revision down
      And the new revision should become the live revision.

  Scenario: Service is deprecated
    Given that I have a piece of content that represents a service
      And that service is being shelved in two weeks
      And that no other services are being shelved
    When I want my service's details removed from the live db
    Then I should be able to choose the date that service is sunsetted
      And that date should not affect other services
```

**Comment by Snugug** [comment]
@scottnath The user being messaged would be a `Content Reviewer` not an application developer. This story is going to have 2 features at least.

If a piece of content doesn't have a shared sunset date, how do you sunset content?

**Comment by Snugug** [comment]
Can you please elaborate on how you envision a piece of content being sunset if each revision has its own sunrise and sunset date? What happens if a revision is sunset? Does it revert to the previous revision? How do you take content down entirely? What does it mean to have a piece of content with three sunset dates?

**Comment by scottnath** [comment]
Hmm...maybe there is a third (or possibly fourth) date here somewhere?

`Content Sunrise`
`Content Sunset`
(as opposed to `revision sunrise`/`revision sunset`)

## Use of multiple sunsets:

Rev1:
Up through 2017
Rev2:
Up from mid-sept->Nov 1 (Halloween Sale!)
Rev3:
Nov 1->Dec 24 (Christmas Sale!)

**Comment by scottnath** [comment]
Then you have logic for which is live...if a later approval, then that content is king if there is a revision date conflict...

There could also be an option to change a specific revision's sunrise/set _without_ creating a new revision but instead updating the existing revision.

**Comment by Snugug** [comment]
That same thing can be achieved by setting the sunrise for Rev 1 to be Dec 25 and having a single sunset date shared between all revisions, and sunrising the revisions when needed. 

~ Sam Richard
(775) 476-8484
http://snug.ug

**Comment by scottnath** [comment]
Not really. What if Rev1 was in there long before Rev2 or Rev3?
Rev1 in this situation is the constant, where subsequent revisions are temporary changes that should come up and go down as needed.

**Comment by Snugug** [comment]
Rev 1 can still be re-submitted with a new sunrise date

It feels very much like you're talking about an edge case that is going to add a tonne of complexity and work to this. I don't see a usecase where re-submitting a revision with a new sunrise can't solve the same thing having separate content and revision sunrise/set dates have.

Actually, I'm concerned having them act independently may create more unintended side-effects than it would resolve. What happens if revision sunrise/sunset blocks overlap? What happens if a revision is published to the "constant" (which there is no such thing as a baseline for content) and scheduled to go live during a window where other revisions are in sunrise/sunset? Who wins? I think this is literally the definition of compelcting; interweaving things together.

**Comment by scottnath** [comment]
Actually, I feel the complexity comes from removing the availability of sunset from the revision level.

I think the latest approval should win the who's-live wars - because you can always add a new revision with the dates you want to win.

For instance:

Rev1:
Up forever
Rev2:
Up from mid-sept->Oct 31 (Halloween Sale!)
Rev3:
Nov 12->Dec 24 (Christmas Sale!)

## Without sunset available at revision:

Rev 1 (sunrise Aug 2016)
Rev 2 (price change, sunrise Sept 20)
Rev 3 (original again, sunrise Nov 1)
Rev 4 (price change, sunrise Nov 12)
Rev 4 (original again again, sunrise Dec 25)

## With sunset

Rev 1 (Aug '16)
Rev 2 (9/20-10/31)
Rev 3 (11/12-12/24)

**Comment by Snugug** [comment]
For _that specific usecase_, what about all of the other ones?

**Comment by scottnath** [comment]
Well, rather than the specifics of a pricing use case (although pricing variations are hardly an edge case,) let's look at the overall direction of content.

Without revision-sunset, content can only move forward, being replaced with the next revision with a sunrise. No temporary changes could happen. If temporary changes are desired, a revision has to be made to add the temp change, then a new revision added to remove it.

With revision-sunset, it can move in both directions without a large planning effort by the content creator.

Also...from a much earlier comment you said `Allow only a single sunset date per content type` - is that a typo or do you mean every piece of content in a content type must share the same sunset date? Shouldn't this be `Allow only a single sunset date for each piece of content`?

**Comment by Snugug** [comment]
Yah, meant `per piece of content`, not content type

Researching how other CMSes handle this, every single one does it for the conceptual piece of content, not a given revision of content. While I know that we are specifically designing Punchcard differently, I'm concerned about the heavy interweaving of revisions and keeping track of that in one's head. Only allowing content to move forward also matches our revision flow, where edits to content are not allowed, only new revisions. Immutable.

**Comment by Snugug** [comment]
From Jeremy Hodge, from a quick interview:

```
* Sunrise/Sunset Date in a revision is for the piece of content
  * A new revision would override
  * All dates would apply to the piece of content
  * Approver modifies sunrise/sunset date, for all revisions
      * Approver can modify sunrise/sunset date
* Rev A published, Rev B being reviewed
    * Sunrise/Sunset of B overrides A
    * Sunset applies to Revision A as well
    * Sunrise can live independently per revision, sunset is shared
* Pricing
    * Rev A - Current Price
    * Rev B - Temp Price
    * Set Rev A set sunrise/set Oct-Dec, Rev B for special pricing
        * Rev B set sunrise Dec 20
        * Rev B sunset Dec 25
* Seems confusing to have a separate content sunrise/sunset
    * Done in past at the content level
    * Set sunrise/sunset to components and the overall page
    * At the page level, set the sunset for 2026 so wouldn't get deleted
    * Components you want to be specific about
* Concern about having revisions with individual dates and a "master" one that needs to be dealt with would be too much
    * Given current options, would prefer the additive method
```

He also hasn't played around with the revisions model we have in place, so he was going to report back after he looked

**Comment by scottnath** [comment]
I guess I'm not seeing this as such a complicated thing programatically. The one with the latest approval and within sunrise/sunset will always win, no? The date of approval will trump all. That way, content can always be replaced by creating/approving a revision and having comparable dates to the live content,
And yet, that would still give granular control over what will go-up/come-down and when, without having to keep creating new revisions for each date if you have temporary changes.

**Comment by scottnath** [comment]
`Concern about having revisions with individual dates and a "master" one`
"master" would become the latest approved with current dates
* would then have an indication in UI for `live` revision
  * could also be prominent in UI (`edit live revision`)

There is some logic to deal with here for sure, but there are also clear use cases for temporary revisions to be live.

**Comment by Snugug** [comment]
There are two things to consider here, the complexity of the implementation, and more important, the user expectations.

w/r/t complexity of implementation, it's more complicated than I think you believe. We would need to have fall-through, which means constantly monitoring all sunrise/sunset dates and juggling them around. Having a single sunset date and as many sunrise as you'd like allows us to "set and forget" the sunrise, which is a much easier (and less error prone) implementation.

The user expectation thing, though, I think is the more important piece. From the conversation with Jeremy, he expressed in the general sense having sunrise/sunset behave as I suggested, but when presented with the pricing use case you suggest, recommended your proposed solution. After realizing he wanted to opposing things, and considering the options, he said he would prefer the additive method. This is one data point, but it's from an actual content editor.

**Comment by Snugug** [comment]
I'm throwing `blocked` on this until I can get a more clear idea of direction for sunset

**Comment by Snugug** [comment]
I'm also going to be doing some user research to get a more clear understanding of what the expectations are

**Comment by scottnath** [comment]
Excellent, best approach (blocked/research)
Ya got my 2 cents!

**Comment by Snugug** [comment]
More research, this time from @bobbinrobyn:

```
* Sunrise/Sunset
  * Choose date and time content to appear/dissapear
* Rev A is live, publish new changes Rev B
  * On Rev B's sunrise, it would "overtake" Rev A live
  * What happens if you put sunrise and sunset for B
    * don't specify what to put up in its place
    * maybe one or the other?
* If Sunset is set
  * On date and time, that content goes away
  * Rev A is live until August 17th (sunset)
  * Rev B sunrise on August 16th w/no sunset
    * Rev B would override Rev A's sunset
* Pricing
  * Rev A - standard pricing
  * Rev B - holiday pricing
  * Rev A - No sunrise or sunset, Rev B has both sunrise and sunset
    * Set Rev B to show up Dec 1
    * Rev B sunset on Dec 31
    * Expect to go back to Rev A after sunset
* Having both content and revision sunrise/sunset sounds confusing in practice although understood in theory
* Leaning towards additive
  * Want something that's telling her what revision is live, what is scheduled (and for when)
* Third option:
  * Have different states for content
  * Custom input plugin
* Not totally opposed to either way
```

**Comment by Snugug** [comment]
From Hannah

```
* Sunrise controls when that piece of content goes up and sunset when it comes down
  * This revision
  * Rev A - currently live, Rev B - something to be published
    * Rev B will replace Rev A for sunrise when published
      * Sunset revert to revision A if it's past Revision B -> Fall Through
      * Sunset moot for Rev A and Rev B's sunset overrides -> Addition
        * Make a new revision Rev C to change sunset
  * Fall through may be more preferred as you'll have a "backup"
    * Would need to set a sunset for the entire content, "makes whole system fail"
    * Both Content and Revision Sunrise/Sunset
      * Separate sunrise/sunset seems redundant or inconsistent
        - Define at beginning
        - Continue Revising
        - Decide want to change? Context has potentially changed too much
      * Would you need to set content sunrise/sunset for every revision?
        * Won't have a standard understanding of how the two are related, especially as they start to change
  * Addition
    * Set a sunset but don't want content to disappear and forget to set a new revision in its place
* Need an approval workflow view of some sort
  * See revisions in order they'd be put live with their sunrise/sunset dates
    * Precedence of each revision in the system
    * Especially for addition. If not lined up correctly, causing problems for yourself
* Approval Process
  * Instead of single screen "Send", able to send revision back through approval process as long as it's not currently waiting to be approved.
```

Hannah is interesting. She prefers the fall-through method (different than both Jeremy and Robyn) but also has identified more potential pitfalls with the fall-through method than the addition method.

**Make sure that Sunrise/Sunset is available in Approval view**

**Comment by SirSpidey** [comment]
@Snugug: To help with granularity, what does it mean that sunrise/sunset it as the content level instead of content type? How do you identify that granularity of content with SR/SS? I've seen SR/SS at content type level only so far.

**Comment by Snugug** [comment]
Each piece of content, say the Conversation API, controls its own SR/SS. All conversation we've had so far is at that granularity as opposed to the "API Content Type" level.

**Comment by Snugug** [comment]
More Research, from @byronhulcher

```
* Sunrise/Sunset
  * Sunrise the date the content is publically accesisble
  * Sunset is the date when the content is no longer publically accessible
  * Initially though sunruse/sunset was universal for a piece of content
* Rev A - Currenty Live, Rev B - In review, ready to be published
  * Sunrise
    * At sunrise date for B, would overtake Rev B for currently live version
  * Sunset
    * As a stack, after sunset, went back to Rev A
    * If Rev A doesn't have sunset, would remove Rev B
    * How to sunset A?
      * Separate sunset for content Foo
      * And all revisions would be past sunset date
      * Set Revision/Content Sunrise/Sunset separately.
* /\ -> Fall Through
  * Planning on having different parts of the content that go live at different times
* Addition - Sunrise for revision, sunset for content (shared)
  * Assumes revisions aren't temporary
  * "Those are the edits, they stay there forever"
* In general case, which makes most sense?
  * Addition seems more straight forward for like Wordpress-style content
  * Fall Through for more complicated cases
* Pricing
  * Fall Through
  * Custom Input Plugin (for this special case*)
    * Doesn't work at scale, ~ >3

* Addition is easiest to reason about
  * Rev A->B->C (=== A)
  * w/Fall Through, in Git, a revert commit, but it's still a commit moving forward
```

**Comment by Snugug** [comment]
Last bit of research for the day, from @toddself

```
* Works with CouchDB a lot
* Revisions only ever increase
  * Two may be identical w/different inbetween
  * Always move forward, never go back
  * Make cloning to `head` state easy, latest index
    * Never confusion as to what version of the content being looked at
* Has worked on ~5 CMSes
* Previous position w/Conde Naste

* Sunrise / Sunset
  * Sunrise
    * Date and time content available in the publically available API
  * Sunset
    * Date after which the content is no longer available in the API
* Rev A -> live content, Rev B -> Content awaiting approval
  * Given Rev A is live with no sunset date and Rev B has a sunset date
    * Rev B will replace Rev A in the API
    * Rev B is the version being sent out
    * Rev B's sunset date takes prescedence
      * Sunsets the piece of content
      * Moving forward, once Rev B is the main version, Rev A is a historical footnote in CMS
  * Given Rev A has a sunset date and Rev B does not
    * Rev B would flag as conflict
    * Either accept Rev A's sunset date or that sunset date will dissapear w/my revision
    * Most recently live revision of the content is the one that own sunset. -> Addition
* Pricing
  * Rev A -> Live content, Rev B -> Awaiting approval
    * Rev B is temporary, constrainted to a given timeframe (think holiday pricing)
    * Rev A live until Rev B goes live, then Rev B is sunset and falls back to Rev A
      * Head always most accurate version of content
      * Make new revision C with new sunrise/sunset content
      * "Spends lots of time in Git"
      * Never go backwards in time
      * Don't just pop off old revision
      * Take revision and copy it in to new head state.
      * Explicitly so no fallback or roll back
      * Always make a new version with `head` piece of content
```

Also had the same concerns that I raised earlier about complexity of fall through content.


With these five bits of research, and a pretty unanimous opinion of the individuals, I think the way to move forward tomorrow is with the `addition` method that I presented. First thing tomorrow I'll update the story itself to reflect this with new features and scenarios as needed then get to work.

Unblocking

**Comment by Snugug** [comment]
I've updated the _crap_ out of this story, and it's now gotten quite a bit larger. Will start working through this this afternoon.

While this is larger, I do not believe it is quite epic sized

**Comment by scottnath** [comment]
@Snugug this is a very clear story. 

One change: I feel that `View Revision Schedule` should be it's own story - large task and includes a new interface.

**Comment by Snugug** [comment]
While I agree its a new interface, it'll mostly wind up being a view of a new table I'm going to be creating, so I'm not concerned about its size, and it's fairly critical to understanding sunrise/sunset. It'll be the last thing I work on, and if I feel it'll wind up being too much I'll divide it up

**Comment by joshkimmell** [comment]
@Snugug please move this into a milestone if you're going to work on it.

**Comment by joshkimmell** [comment]
@Snugug: Is this going to be done in this milestone?

**Comment by Snugug** [comment]
Split out the approver view to #354 as you suggested @scottnath

**Comment by Snugug** [comment]
Also sized that and added it to this sprint; I'll work on that once this is in
